src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Array.cs

```diff
@@ -0,0 +1,272 @@
+            Assert.Equal(expectedLength, (int)length!.Value);
+
+            for (int i = 0; i < actualLength; i++)
+            {
+                reader.ReadInt64();
+            }
+
+            Assert.Throws<FormatException>(() => reader.ReadInt64());
+        }
+
+        [Theory]
+        [InlineData("81", 1, 0)]
+        [InlineData("828101", 2, 1)]
+        [InlineData("8681018102", 6, 2)]
+        public static void ReadArray_IncorrectDefiniteLength_NestedValues_ShouldThrowFormatException(string hexEncoding, int expectedLength, int actualLength)
+        {
+            byte[] encoding = hexEncoding.HexToByteArray();
+            var reader = new CborReader(encoding);
+
+            ulong? length = reader.BeginReadArray();
+            Assert.Equal(expectedLength, (int)length!.Value);
+
+            for (int i = 0; i < actualLength; i++)
+            {
+                ulong? innerLength = reader.BeginReadArray();
+                Assert.Equal(1, (int)innerLength!.Value);
+                reader.ReadInt64();
+                reader.EndReadArray();
+            }
+
+            Assert.Throws<FormatException>(() => reader.ReadInt64());
+        }
+
+        [Fact]
+        public static void BeginReadArray_EmptyBuffer_ShouldThrowFormatException()
+        {
+            byte[] encoding = Array.Empty<byte>();
+            var reader = new CborReader(encoding);
+
+            Assert.Throws<FormatException>(() => reader.BeginReadArray());
+        }
+
+        [Theory]
+        [InlineData("00")] // 0
+        [InlineData("20")] // -1
+        [InlineData("40")] // empty byte string
+        [InlineData("60")] // empty text string
+        [InlineData("f6")] // null
+        [InlineData("a0")] // {}
+        [InlineData("f97e00")] // NaN
+        [InlineData("fb3ff199999999999a")] // 1.1
+        public static void BeginReadArray_InvalidType_ShouldThrowInvalidOperationException(string hexEncoding)
+        {
+            byte[] data = hexEncoding.HexToByteArray();
+            var reader = new CborReader(data);
+            Assert.Throws<InvalidOperationException>(() => reader.BeginReadArray());
+        }
+
+        [Theory]
+        // Invalid initial bytes with array major type
+        [InlineData("9c")]
+        [InlineData("9d")]
+        [InlineData("9e")]
+        // valid initial bytes missing required definite length data
+        [InlineData("98")]
+        [InlineData("9912")]
+        [InlineData("9a000000")]
+        [InlineData("9b00000000000000")]
+        public static void BeginReadArray_InvalidData_ShouldThrowFormatException(string hexEncoding)
+        {
+            byte[] data = hexEncoding.HexToByteArray();
+            var reader = new CborReader(data);
+
+            Assert.Throws<FormatException>(() => reader.BeginReadArray());
+        }
+    }
+
+    static class ArrayReaderHelper
+    {
+        public static void ReadArray(this CborReader reader, params object[] expectedValues)
```

Comment:
The first test used this, and I was trying to figure out what the semantics on the reader were.

This should definitely be Verify instead of Read, and I'd recommend not making an extension method (because it makes the tests not look like real code).

```suggestion
        public static void VerifyArray(CborReader reader, params object[] expectedValues)
```